File: ruby_vra8_plus/eg.rb
require './common'
require './env_table'

class EG
  STATE_ATTACK = 0
  STATE_DECAY = 1
  STATE_SUSTAIN = 2
  STATE_RELEASE = 3
  STATE_IDLE = 4

  def initialize
    @attack_speed = 255
    @decay_speed = 255
    @sustain_level = 127
    @state = STATE_IDLE
    @count = 0
    @level = 0
  end

  def set_attack_time(attack_time)
    @attack_speed = $env_table_speed_from_time[attack_time]
  end

  def set_decay_time(decay_time)
    @decay_speed = $env_table_speed_from_time[decay_time]
  end

  def set_sustain_level(sustain_level)
    @sustain_level = sustain_level
  end

  def note_on
    if (@level == 127)
      @state = STATE_DECAY
      @count = 0
    else
      @state = STATE_ATTACK
      @count = $env_table_attack_inverse[@level] * 256
    end
  end

  def note_off
    case (@state)
    when STATE_ATTACK, STATE_DECAY, STATE_SUSTAIN
      @state = STATE_RELEASE
      @count = $env_table_decay_inverse[@level] * 256
    end
  end

  def sound_off
    @state = STATE_IDLE
    @count = 0
    @level = 0
  end

  def clock
    case (@state)
    when STATE_ATTACK
      @count += @attack_speed
      if ((@count / 256) < 255)
        @level = $env_table_attack[@count / 256]
      else
        @state = STATE_DECAY
        @count = 0
        @level = 127
      end
    when STATE_DECAY
      @count += @decay_speed
      @level = $env_table_decay[@count / 256]
      if (@level <= @sustain_level)
        @state = STATE_SUSTAIN
        @level = @sustain_level
      end
    when STATE_SUSTAIN
      @level = @sustain_level
    when STATE_RELEASE
      @count += @decay_speed
      if ((@count / 256) < 255)
        @level = $env_table_decay[@count / 256]
      else
        @state = STATE_IDLE
        @count = 0
        @level = 0
      end
    when STATE_IDLE
      @level = 0
    end

    return @level
  end
end

File: eg.js
var EG = function() {
  const STATE_ATTACK        = 0;
  const STATE_DECAY_SUSTAIN = 1;
  const STATE_RELEASE       = 3;
  const STATE_IDLE          = 4;

  this.setAttackTime = function(attackTime) {
    this.attackTime = attackTime;
  };

  this.setDecayTime = function(decayTime) {
    this.decayTime = decayTime;
  };

  this.setSustainLevel = function(sustainLevel) {
    this.sustainLevel = sustainLevel;
  };

  this.setReleaseTime = function(releaseTime) {
    this.releaseTime = releaseTime;
  };

  this.noteOn = function() {
    this.state = STATE_ATTACK;
  };

  this.noteOff = function() {
    this.state = STATE_RELEASE;
  };

  this.soundOff = function() {
    this.state = STATE_IDLE;
  };

  this.clock = function() {
    var as = 10 / Math.pow(10, (127 - this.attackTime) / (127 / 3));
    var ar = Math.pow(1 / 3, 1 / (SAMPLING_RATE * as));
    var ds = 10 / Math.pow(10, (127 - this.decayTime) / (127 / 3));
    var dr = Math.pow(1 / 32, 1 / (SAMPLING_RATE * ds));
    var sl = this.sustainLevel / 127;
    var rs = 10 / Math.pow(10, (127 - this.releaseTime) / (127 / 3));
    var rr = Math.pow(1 / 32, 1 / (SAMPLING_RATE * rs));

    switch (this.state) {
    case STATE_ATTACK:
      this.level = 1.5 - ((1.5 - this.level) * ar);
      if (this.level >= 1) {
        this.state = STATE_DECAY_SUSTAIN;
        this.level = 1;
      }
      break;
    case STATE_DECAY_SUSTAIN:
      if (this.level > sl) {
        if ((this.level - sl) <= (1 / 1024)) {
          this.level = sl;
        } else {
          this.level = sl - ((sl - this.level) * dr);
        }
      }
      break;
    case STATE_RELEASE:
      this.level = 0 - ((0 - this.level) * rr);
      if (this.level <= 1 / 1024) {
        this.state = STATE_IDLE;
        this.level = 0;
      }
      break;
    case STATE_IDLE:
      this.level = 0;
      break;
    }

    return this.level;
  };

  this.attackTime   = 0;
  this.decayTime    = 0;
  this.sustainLevel = 127;
  this.level        = 0;
  this.releaseTime  = 0;
};
